src/rulesets/Base/array.jl

```diff
@@ -527,3 +527,42 @@ function _extrema_dims(x, dims)
     end
     return y, extrema_pullback_dims
 end
+
+#####
+##### `pmap`
+#####
+
+using UUIDs # of course we don't want this;
+            # what is a better way of uniquely identifying function calls from the forward pass?
+
+const save_backs = Dict() # could this be a stack instead?
```

Comment:
Does it matter that a Dict isn't thread safe?

If not, then the obvious way to label entries is just to increment `const COUNTER = Ref(1)`.

Reply:
Took a stab at this!

Reply:
I do worry about the thread-safe issue though. Perhaps it would be best to explicitly restrict to `batch_size = 1`, and port over the Zygote rule to run as a backup for `batch_size > 1`?

Reply:
You could surely make it thread-safe, via Atomic & all that. But maybe this becomes slow?

I guess I still don't know why one global Dict is better than one Array... which can be created within the rule & then everyone knows where to write. Why is a Dict better? How do you test that this is working?

Reply:
I... don't know why I used a global Dict. I can't see any issues with declaring something within the `rrule`. I'll give that a try now.

Reply:
I guess it makes initializing the array on the processors a little more complex -- do I want to initialize it on every processor in the worker pool before the `pmap`, or initialize it inside of `forw` if it isn't already defined?

Reply:
Oh, I think I see why I used a global dictionary. My understanding is that the worker processors don't know anything about the scoping on the root processor. So, to persist state on a worker process, the simplest solution is to just have a global variable. Does that make sense?